### ISSUE TITLE
(`c2rust-analyze`) Handle self-referential/recursive `Trivial` types

### ISSUE THREAD

**Issue by kkysen** [immunant/c2rust#834]
#817 implements skipping `Trivial` function calls, but it doesn't yet handle self-referential/recursive types, such as something like these:

```rust
struct A<'a> {
    a: &'a A,
}
```

Currently (in #817), this would infinitely recurse.

Rather than implementing this all in #817, I'll split the fix into a separate PR to reduce each PR's complexity, as #817 is still a significant improvement in handling many more function calls.

See:
* https://github.com/immunant/c2rust/pull/817#issuecomment-1421205333
* https://github.com/immunant/c2rust/pull/817#issuecomment-1424582551
* https://github.com/immunant/c2rust/pull/817#discussion_r1101951279
* https://github.com/immunant/c2rust/pull/817#discussion_r1101952287
* https://github.com/immunant/c2rust/pull/817#discussion_r1102081667

**Comment by spernsteiner** [comment]
Another case to consider is pathologically large types, such as this:
```Rust
type T1 = ();
type T2 = (T1, T1);
type T3 = (T2, T2);
// ...
type T99 = (T98, T98);
```
Checking triviality of `T99` will run forever (approximately), without running out of memory or overflowing the stack.

I think the same kind of caching I describe in the comments linked above would also work for types like `T99`.

**Comment by kkysen** [comment]
I don't think we have to deal with pathologically large ZSTs yet, not before self-referential types.  At least the fix for them is the same for the plausible case.
